rllib/core/models/base.py

```diff
@@ -0,0 +1,239 @@
+import abc
+from dataclasses import dataclass
+from typing import List, Union
+
+from ray.rllib import SampleBatch
+from ray.rllib.models.specs.specs_dict import SpecDict
+from ray.rllib.models.specs.specs_base import DummySpec
+
+from ray.rllib.utils.annotations import ExperimentalAPI
+from ray.rllib.utils.typing import TensorType
+from ray.rllib.utils.nested_dict import NestedDict
+
+
+# Top level keys that unify model i/o.
+STATE_IN: str = "state_in"
+STATE_OUT: str = "state_out"
+ENCODER_OUT: str = "encoder_out"
+# For Actor-Critic algorithms, these signify data related to the actor and critic
+ACTOR: str = "actor"
+CRITIC: str = "critic"
+
+
+@ExperimentalAPI
+@dataclass
+class ModelConfig(abc.ABC):
+    """Base class for model configurations.
+
+    ModelConfigs are framework-agnostic.
+    A ModelConfig is usually built by RLModules by querying it form a Catalog object.
+    It is therefore a means of configuration for RLModules.
+    However, that is not a limitation, and they can be built directly as well.
+
+    Attributes:
+        output_dim: The output dimension of the network.
+
+    Usage Example together with a Model:
+
+    .. testcode::
+
+        class MyModel(Model):
+            def __init__(self, config):
+                super().__init__(config)
+                self.my_param = config.my_param * 2
+
+            def _forward(self, input_dict):
+                return input_dict["obs"] * self.my_param
+
+
+        @dataclass
+        class MyModelConfig(ModelConfig):
+            my_param: int = 42
+
+            def build(self, framework: str):
+                if framework == "bork":
+                    return MyModel(self)
+
+
+        config = MyModelConfig(my_param=3)
+        model = config.build(framework="bork")
+        print(model._forward({"obs": 1}))
+
+    .. testoutput::
+
+        6
+
+    """
+
+    output_dim: int = None
+
+    @abc.abstractmethod
+    def build(self, framework: str):
+        """Builds the model.
+
+        Args:
+            framework: The framework to use for building the model.
+        """
+        raise NotImplementedError
+
+
+class Model(abc.ABC):
+    """Framework-agnostic base class for RLlib models.
+
+    Models are low-level neural network components that offer input- and
+    output-specification, a forward method, and a get_initial_state method. Models
+    are composed in RLModules.
+
+    Usage Example together with ModelConfig:
```

Comment:
This is obviously the same example as the above.
This makes sense because the classes are very tightly coupled (so ought to be used together).
Still, both should contain an example in their docstring.